tablator.1
.TH TABLATOR 1 "January 26, 2025" "Game Tools"
.SH NAME
tablator \- prints a random item from a table
.SH SYNOPSIS
.nf
.B tablator [ -d \fIDATA_DIR\fR | --data-dir \fIDATA_DIR\fR ] [ -n \fINUMBER\fR | --number \fINUMBER\fR ]
     [ -l | --list ] [ -p | --print ] [ -t | --trace ] [ -v | --verbose ]
     [ \fITABLE1\fR \fITABLE2\fR ... ]
.B tablator [-d \fIDATA_DIR\fR ] [ -l | --list ]
.B tablator [ -p | --print ] table
.B tablator [ -h | --help ]
.fi
.SH DESCRIPTION
.PP
Randomly select items from lookup tables.
.PP
There are two parts to this project, the tablator (Python) module and the
command line tool.  This page describes the command line tool.
.PP
The \fBtablator\fR tool is both an example of how to use the library and allows
use of the data sets.  It lists the available tables, prints tables to the
console, and roll (select a random row) on those tables.  The tools requires a
data directory, defined directly as a program argument, in the environment,
or it falls back to the current working directory.
.PP
The \fBtablator\fR Python module is a general purpose library for randomly
selecting results from look-up tables.  Each row in the table has a weight.
A number is randomly generated from 1 to "total weight" and the row
corresponding to that number is selected.
.PP
Tab-LA-tor. Originally called "treasure tables" was developed to roll up
magic items from the treasure tables in the AD&D DMG.
The name is a compaction of "table-ator".
It has nothing to do with guitar tablature.
.SH OPTIONS
.TP
.BR \-h ", " \-\-help
Show usage information and exit
.TP
.BR \-d " " \fIDATA_DIR\fR ", " \-\-data-dir " " \fIDATA_DIR\fR
Set the table data directory
.TP
.BR \-n " " \fINUMBER\fR ", " \-\-number " " \fINUMBER\fR
Number of items to select from each table (default 1)
.TP
.BR \-l ", " \-\-list
List available tables (table file names sans extension in the data directory)
.TP
.BR \-p ", " \-\-print
Print the table (plain text)
.TP
.BR \-t ", " \-\-trace
Enable trace messages
.TP
.BR \-v ", " \-\-verbose
Enable debug messages
.SH EXIT STATUS
Zero if successful, 1 on any error.
.SH ENVIRONMENT
.BR TABLATOR_DATA_DIR
The data directory.
.SH EXAMPLES
.nf
python3 tablator.py --help
python3 tablator.py --data-dir ~/data/add-dmg --list
export TABLATOR_DATA_DIR=~/data/add-dmg
python3 tablator.py lair-a
python3 tablator.py --number 3 potions
python3 tablator.py --print scrolls
python3 tablator.py --data-dir ~/data/pathfinder-encounters dungeon-mid
python3 tablator.py magic-item -n 3
python3 tablator.py weapons rings scrolls
.fi
.SH BUGS
Various.
.SH SEE ALSO
The \fBREADME.md\fR file in
.UR https://github.com/crithead/tablator
.UE
along with the source.



src/tablator.py
#!/usr/bin/env python3
"""
Treasure Table -- Generate
"""

import argparse
import os
import sys
import traceback

import tablator
import tablator.logger


def apply_environ(args):
    """
    Fill in missing argument values from the environement.
    Options: data_dir
    """
    if args.trace: print('=== apply_environ')

    if args.data_dir is None and 'TABLATOR_DATA_DIR' in os.environ:
        args.data_dir = os.environ['TABLATOR_DATA_DIR']
        if args.verbose:
            print(f'--- Found TABLATOR_DATA_DIR = {args.data_dir}')


def check_data_dir(args):
    """
    Check DATA_DIR.
    Set to PWD if None.
    Riase ValeError is non-existent.
    """
    tablator.logger.trace('check_data_dir')

    if args.data_dir is None:
         args.data_dir = os.getcwd()

    if os.path.isdir(args.data_dir):
        tablator.logger.debug(f'Using DATA_DIR = {args.data_dir}')
    else:
        raise ValueError(f'DATA DIR ({args.data_dir}) does not exist')


def get_args():
    """
    Get command line arguments.
    """
    parser = argparse.ArgumentParser(
        description='Roll on a table',
        epilog='\n')
    parser.add_argument('-d', '--data-dir', action='store', default=None,
                        help='let the table data directory')
    parser.add_argument('-l', '--list', action='store_true', default=False,
                        help='list available tables')
    parser.add_argument('-n', '--number', action='store', default=1, type=int,
                        help='number of rolls on each table')
    parser.add_argument('-p', '--print', action='store_true', default=False,
                        help='print the table (plain text)')
    parser.add_argument('-t', '--trace', action='store_true', default=False,
                        help='enable trace messages')
    parser.add_argument('-v', '--verbose', action='store_true', default=False,
                        help='enable debug messages')
    parser.add_argument('tables', nargs='*',
                        help='table names')
    return parser.parse_args()


if __name__ == '__main__':
    exit_code = None
    try:
        args = get_args()
        apply_environ(args)
        tablator.logger.set(args.verbose, args.trace)

        check_data_dir(args)
        tablator.set_data_dir(args.data_dir)

        if args.list:
            table_list = tablator.data.list_tables()
            print('', 'Available tables', '', sep='\n')
            for table_name in sorted(table_list):
                print('   ', table_name)
            print()
            exit_code = 0

        else:  # do table lookups
            exit_code = 1
            for table_name in args.tables:
                if args.print:
                    tablator.print_plain(table_name)
                else:
                    items = tablator.generate(table_name, args.number)
                    for item in items:
                        print(item)
                exit_code = 0

    except ValueError as e:
        traceback.print_exc()
        print(repr(e), file=sys.stderr)
        exit_code = 1

    except Exception:
        traceback.print_exc()
        exit_code = 1

    finally:
        if exit_code is None:
            print('Warning: exit_code is None', file=sys.stderr)
            exit_code = 99
        exit(exit_code)
